rllib/env/single_agent_episode.py

```diff
@@ -294,160 +331,265 @@ def is_done(self) -> bool:
         time.
 
+        Args:
+            overlap: The number of timesteps to take into the new chunk, simply for
+                horizon/visibility reasons (but without actually being part of the
+                new chunk). For example, if this episode ends in actions 5, 6, 7,
+                and 8, and we call `cut(overlap=2)`, the returned chunk will have
+                actions 7 and 8 already in it, but still t_started==t==8 (not 7!).
+                Episodes being "finalized" into batches for model forward passing or
+                used for metrics computations should ignore this extra data at the
+                beginning.
+
         Returns:
             The successor Episode chunk of this one with the same ID and state and the
             only observation being the last observation in self.
         """
         assert not self.is_done
 
+        indices_obs_and_infos = slice(-overlap - 1, None)
+        indices_rest = slice(-overlap, None) if overlap > 0 else slice(None, 0)
+
         return SingleAgentEpisode(
             # Same ID.
             id_=self.id_,
-            # First (and only) observation of successor is this episode's last obs.
-            observations=[self.observations[-1]],
-            # First (and only) info of successor is this episode's last info.
-            infos=[self.infos[-1]],
-            # Same state.
-            states=self.states,
+            observations=self.get_observations(indices=indices_obs_and_infos),
+            infos=self.get_infos(indices=indices_obs_and_infos),
+            actions=self.get_actions(indices=indices_rest),
+            rewards=self.get_rewards(indices=indices_rest),
+            extra_model_outputs={
+                k: self.get_extra_model_outputs(k, indices_rest)
+                for k in self.extra_model_outputs.keys()
+            },
             # Continue with self's current timestep.
             t_started=self.t,
         )
 
-    def to_sample_batch(self) -> SampleBatch:
-        """Converts a `SingleAgentEpisode` into a `SampleBatch`.
+    def split_at(self, index: int) -> "SingleAgentEpisode":
+        """Splits at given index, keeping the 1st half in `self` and returning 2nd half.
+
+        Changes `self` in-place by making it shorter (from 0 to [original length] to
+        0 to `split_at`). The second "half" Episode from split_at till the end is
+        returned as a new SingleAgentEpisode object.
 
-        Note that `RLlib` is relying in training on the `SampleBatch`  class and
-        therefore episodes have to be converted to this format before training can
-        start.
+        Args:
+            index: The index at which to perform the split.
 
         Returns:
-            An `ray.rLlib.policy.sample_batch.SampleBatch` instance containing this
-            episode's data.
+            The new SingleAgentEpisode representing the second half of the split. The
```

Comment:
Isn't it better to return a new episode for the first half?

Why don't you use slicing API episode[t1:t2]? then slice(t1) would be episode[t1:]

Reply:
Great idea! Fixed and test case adjusted and enhanced.

Reply:
Both `slice([some slice object])` and `episode[t0:t1]` (squared bracket syntax) are now supported.